app/views/govuk_component/task_list.raw.html.erb

```diff
@@ -17,70 +19,132 @@
+              <% if step_index > 0 and logic.length %>
+                <div class="pub-c-task-list__circle pub-c-task-list__circle--logic">
+                  <span class="pub-c-task-list__circle-inner">
+                    <span class="pub-c-task-list__circle-background">
+                      <%= logic %>
+                    </span>
+                  </span>
+                </div>
+              <% end %>
+
               <div class="pub-c-task-list__header js-toggle-panel" data-position="<%= "#{group_index + 1}.#{step_index + 1}" %>">
                 <h<%= heading_level %> class="pub-c-task-list__title js-step-title">
                   <%= step[:title] %>
                 </h<%= heading_level %>>
               </div>
 
               <div class="pub-c-task-list__panel js-panel" id="step-panel-<%= id %>-<%= step_index + 1 %>">
-                <div class="pub-c-task-list__panel-content">
-                  <%= step[:panel].html_safe if step[:panel] %>
+                <%
+                  in_substep = false
+                  link_index = 0
+                %>
+                <% step[:contents].each do |element| %>
+                  <% if element[:type] == 'paragraph' %>
+                    <p class="pub-c-task-list__paragraph"><%= element[:text] %></p>
 
-                  <% step[:panel_descriptions].to_a.each do |panel_description| %>
-                    <p class="pub-c-task-list__panel-description">
-                      <%= panel_description %>
-                    </p>
-                  <% end %>
-                </div>
+                  <% elsif element[:type] == 'list' %>
+                    <%
+                      list_style = ''
+                      list_element = 'ol'
+
+                      if element[:style] == 'required'
+                        list_style = 'pub-c-task-list__links--required'
+                      elsif element[:style] == 'choice'
+                        list_style = 'pub-c-task-list__links--choice'
+                        list_element = 'ul'
+                      end
+                    %>
+                    <<%= list_element %>
+                      class="pub-c-task-list__links <%= list_style %>"
+                      data-length="<%= element[:links].length %>"
+                      >
+                      <% element[:links].each do |link| %>
+                        <%
+                          link_text = link[:text]
+
+                          if link[:href]
+                            link_href = link[:href]
+                            link_index += 1
+                          end
+
+                          if link[:active]
+                            link_href = "#content"
+                            link_text = '<span class="visuallyhidden">You are currently viewing:</span> '.html_safe + link_text
+                          end
+                        %>
+                        <li class="pub-c-task-list__link">
+                          <% if link[:href] %>
```

Comment:
There are lots of checks for `link[:href]` here. Would it be simpler to construct the link inside one `link[:href]` if statement, and then simply have an else clause for displaying the text if no href exists? I'm possibly missing some context here.

Reply:
@vanitabarrett @alecgibson I've rewritten this whole section because yes, it was a bit of a mess. Hopefully considerably better now.

Links without a href are used to for the or/then option, where we want a list that contains a link, then the text 'or', then another link. It's easier to put this option in than try to make the or/then option considerably more complex, plus containing those three elements in the same list is a good way to associate them.